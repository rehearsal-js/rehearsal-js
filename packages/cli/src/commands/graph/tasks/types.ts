export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
