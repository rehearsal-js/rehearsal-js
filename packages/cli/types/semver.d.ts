declare module 'semver' {
  const semver: {
    valid: (input: string) => boolean;
  };
  export = semver;
}
