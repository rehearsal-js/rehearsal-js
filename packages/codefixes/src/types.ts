import {
  CodeFixAction,
  DiagnosticWithLocation,
  FileTextChanges,
  LanguageService,
  Node,
  Program,
  TypeChecker,
} from 'typescript';

export interface DiagnosticWithContext extends DiagnosticWithLocation {
  service: LanguageService;
  program: Program;
  checker: TypeChecker;
  node?: Node;
}

export interface CodeHintList {
  [key: number]: CodeHintListElement;
}

export interface CodeHintListElement {
  hint: string;
  helpUrl?: string;
  hints?: CodeHint[];
}

export interface CodeHint {
  when: (n: Node, p: Program, c: TypeChecker) => boolean;
  hint: string;
}

export interface CodeFixCollection {
  getFixForDiagnostic(diagnostic: DiagnosticWithContext): CodeFixAction | undefined;
}

export type CodeFixKind = 'insert' | 'delete' | 'replace';

export interface FixedFile {
  fileName: string;
  newCode?: string;
  oldCode?: string;
  codeFixAction?: CodeFixKind;
  location: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
}

export interface CodeFix {
  getCodeAction: (diagnostic: DiagnosticWithContext) => CodeFixAction | undefined;
}

export function createCodeFixAction(
  fixName: string,
  changes: FileTextChanges[],
  description: string
): CodeFixAction {
  return {
    fixName,
    description,
    changes,
  };
}
