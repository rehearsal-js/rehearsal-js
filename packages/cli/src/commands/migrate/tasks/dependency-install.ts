import { resolve } from 'node:path';
import { existsSync, promises as fs } from 'node:fs';
import { ListrTask } from 'listr2';

import { addDep } from '@rehearsal/utils';
import { PackageJson } from 'type-fest';
import { MigrateCommandContext, MigrateCommandOptions } from '../../../types.js';

export const REQUIRED_DEPENDENCIES = [
  '@types/node',
  '@typescript-eslint/eslint-plugin',
  '@typescript-eslint/parser',
  'eslint',
  'eslint-config-prettier',
  'eslint-plugin-prettier',
  'prettier',
  'typescript',
];

// get the name of dependency from those two format:
// - foo
// - foo@{version}
// Be aware that a package name would start with @, e.g @types/node
function extractDepName(dep: string): string {
  const reg = /^(@?[^@]+)/g;
  const matched = dep.match(reg);
  return matched ? matched[0] : dep;
}

// check if package.json has all required dependecies
// from rehearsal default and user config
export async function shouldRunDepInstallTask(
  options: MigrateCommandOptions,
  context: Partial<MigrateCommandContext> = {}
): Promise<boolean> {
  const { basePath } = options;
  // add extra required dependencies from user config if there is any
  let dependencies: string[] = [];
  let devDependencies = REQUIRED_DEPENDENCIES;
  if (context.userConfig?.hasDependencies) {
    // check if package.json has all the dependencies listed in user config
    dependencies = [...dependencies, ...context.userConfig.dependencies];
    devDependencies = [...devDependencies, ...context.userConfig.devDependencies];
  }

  const packageJsonPath = resolve(basePath, 'package.json');
  if (existsSync(packageJsonPath)) {
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    for (const d of dependencies) {
      if (!packageJson.dependencies || !packageJson.dependencies[extractDepName(d)]) {
        return true;
      }
    }

    for (const d of devDependencies) {
      if (!packageJson.devDependencies || !packageJson.devDependencies[extractDepName(d)]) {
        return true;
      }
    }
  } else {
    // no package.json
    return true;
  }
  return false;
}

export function depInstallTask(
  options: MigrateCommandOptions,
  context?: Partial<MigrateCommandContext>
): ListrTask {
  return {
    title: 'Install dependencies',
    enabled: (): boolean => !options.dryRun,
    skip: async (ctx: MigrateCommandContext): Promise<boolean> =>
      !(await shouldRunDepInstallTask(options, ctx)),
    task: async (ctx: MigrateCommandContext, task): Promise<void> => {
      // If context is provide via external parameter, merge with existed
      if (context) {
        ctx = { ...ctx, ...context };
      }

      let dependencies: string[] = [];
      let devDependencies = REQUIRED_DEPENDENCIES;
      // install custom dependencies
      if (ctx.userConfig?.hasDependencies) {
        task.output = `Install dependencies from config`;
        dependencies = [...dependencies, ...ctx.userConfig.dependencies];
        devDependencies = [...devDependencies, ...ctx.userConfig.devDependencies];
        if (ctx.userConfig?.hasPostInstallHook) {
          task.output = `Run postInstall hook from config`;
          await ctx.userConfig.postInstall();
        }
      }

      const errorMessages: string[] = [];

      // for deps
      if (dependencies.length) {
        try {
          task.output = `Installing dependecies: ${dependencies.join(', ')}`;
          await addDep(dependencies, false, { cwd: options.basePath });
        } catch (e) {
          errorMessages.push(formatErrorMessage(dependencies, false));
        }
      }

      // for devDeps
      if (devDependencies.length) {
        try {
          task.output = `Installing devDependencies: ${devDependencies.join(', ')}`;
          await addDep(devDependencies, true, { cwd: options.basePath });
        } catch (e) {
          errorMessages.push(formatErrorMessage(devDependencies, true));
        }

        if (errorMessages.length) {
          throw new Error(errorMessages.join('\n'));
        }
      }
    },
    // will print and keep what dpe is currently installing at bottom bar
    options: { persistentOutput: true, bottomBar: Infinity },
  };
}

function formatErrorMessage(deps: string[], dev: boolean): string {
  const depType = dev ? 'devDependencies' : 'dependencies';
  return [
    `We ran into an error when installing ${depType}, please install the following as ${depType} and try again.`,
    `${deps.map((d) => `  - ${d}`).join('\n')}`,
  ].join('\n');
}
