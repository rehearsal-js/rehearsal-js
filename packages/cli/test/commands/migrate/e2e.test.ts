import { resolve } from 'node:path';
import { readFileSync, readdirSync, promises as fs } from 'node:fs';
import { readJSONSync, writeJSONSync } from 'fs-extra/esm';
import { setGracefulCleanup, dirSync } from 'tmp';
import { beforeEach, describe, expect, test } from 'vitest';
import { create, getFiles } from '@rehearsal/test-support';
import yaml from 'js-yaml';
import fixturify from 'fixturify';
import { REQUIRED_DEPENDENCIES } from '../../../src/commands/migrate/tasks/dependency-install.js';

import { runBin, prepareTmpDir, cleanOutput } from '../../test-helpers/index.js';
import { CustomConfig, TSConfig } from '../../../src/types.js';
import type { PackageJson } from 'type-fest';

setGracefulCleanup();

describe('migrate - validation', () => {
  let basePath = '';

  beforeEach(() => {
    basePath = prepareTmpDir('initialization');
  });

  test('pass in a clean project', async () => {
    const { stdout } = await runBin('migrate', ['--ci'], {
      cwd: basePath,
    });

    expect(stdout).toContain('Migration Complete');
  });

  test('throw if not in project root with npm/yarn workspaces', async () => {
    const { name: basePath } = dirSync();
    const files = {
      'package.json': JSON.stringify({
        workspaces: ['packages/*'],
      }),
      packages: {
        'package-a': {
          'package.json': JSON.stringify({
            name: 'package-a',
            version: '1.0.0',
          }),
        },
      },
    };
    fixturify.writeSync(basePath, files);

    const { stdout } = await runBin('migrate', ['--ci'], {
      cwd: resolve(basePath, 'packages', 'package-a'),
    });
    expect(stdout).toContain('migrate command needs to be running at project root with workspaces');
  });

  test('not throw if in project root with unrelated npm/yarn workspaces', async () => {
    const { name: basePath } = dirSync();
    const files = {
      'package.json': JSON.stringify({
        workspaces: ['packages/lib/*'],
      }),
      packages: {
        'package-a': {
          'package.json': JSON.stringify({
            name: 'package-a',
            version: '1.0.0',
          }),
        },
      },
    };
    fixturify.writeSync(basePath, files);

    const { stdout: secondRunStdout } = await runBin('migrate', ['--ci'], {
      cwd: resolve(basePath, 'packages', 'package-a'),
    });
    expect(cleanOutput(secondRunStdout, basePath)).toMatchSnapshot();
  });

  test('throw if not in project root with pnpm workspaces', async () => {
    const { name: basePath } = dirSync();
    const files = {
      'package.json': JSON.stringify({
        name: 'foo',
      }),
      'pnpm-lock.yaml': '',
      'pnpm-workspace.yaml': yaml.dump({ packages: ['packages/*'] }),
      packages: {
        'package-a': {
          'package.json': JSON.stringify({
            name: 'package-a',
            version: '1.0.0',
          }),
        },
      },
    };
    fixturify.writeSync(basePath, files);

    const { stdout } = await runBin('migrate', ['--ci'], {
      cwd: resolve(basePath, 'packages', 'package-a'),
    });
    expect(stdout).toContain('migrate command needs to be running at project root with workspaces');
  });

  test('not throw if in project root with unrelated npm/yarn workspaces', async () => {
    const { name: basePath } = dirSync();
    const files = {
      'package.json': JSON.stringify({
        name: 'foo',
      }),
      'pnpm-lock.yaml': '',
      'pnpm-workspace.yaml': yaml.dump({ packages: ['packages/lib/*'] }),
      packages: {
        'package-a': {
          'package.json': JSON.stringify({
            name: 'package-a',
            version: '1.0.0',
          }),
        },
      },
    };
    fixturify.writeSync(basePath, files);

    const { stdout: secondRunStdout } = await runBin('migrate', ['--ci'], {
      cwd: resolve(basePath, 'packages', 'package-a'),
    });
    expect(cleanOutput(secondRunStdout, basePath)).toMatchSnapshot();
  });

  test('relative entrypoint inside project root works', async () => {
    basePath = prepareTmpDir('basic');

    const { stdout } = await runBin('migrate', ['-e', 'foo.js', '--ci'], {
      cwd: basePath,
    });
    expect(cleanOutput(stdout, basePath)).toMatchSnapshot();
  });

  test('absolute entrypoint inside project root works', async () => {
    basePath = prepareTmpDir('basic');

    const { stdout } = await runBin('migrate', ['-e', resolve(basePath, 'foo.js'), '--ci'], {
      cwd: basePath,
    });
    expect(cleanOutput(stdout, basePath)).toMatchSnapshot();
  });

  test('entrypoint outside project root does not work', async () => {
    basePath = prepareTmpDir('basic');

    const { stdout } = await runBin('migrate', ['-e', resolve(__dirname, 'e2e.test.ts'), '--ci'], {
      cwd: basePath,
    });
    expect(stdout).toContain('Could not find entrypoint');
  });
});

describe('migrate: e2e', () => {
  let basePath = '';

  beforeEach(() => {
    basePath = prepareTmpDir('basic');
  });

  test('default migrate command', async () => {
    const { stdout } = await runBin('migrate', ['--ci'], {
      cwd: basePath,
    });

    // summary message
    expect(cleanOutput(stdout, basePath)).toMatchSnapshot();

    // file structures
    const fileList = readdirSync(basePath);
    expect(fileList).toContain('index.ts');
    expect(fileList).toContain('foo.ts');
    expect(fileList).toContain('depends-on-foo.ts');
    expect(fileList).not.toContain('index.js');
    expect(fileList).not.toContain('foo.js');
    expect(fileList).not.toContain('depends-on-foo.js');

    // file contents
    expect(readFileSync(resolve(basePath, 'foo.ts'), { encoding: 'utf-8' })).toMatchSnapshot();
    expect(
      readFileSync(resolve(basePath, 'depends-on-foo.ts'), { encoding: 'utf-8' })
    ).toMatchSnapshot();
    expect(readFileSync(resolve(basePath, 'index.ts'), { encoding: 'utf-8' })).toMatchSnapshot();

    // Dependencies
    const packageJson = JSON.parse(
      await fs.readFile(resolve(basePath, 'package.json'), 'utf-8')
    ) as PackageJson;

    const devDeps = packageJson.devDependencies;
    expect(Object.keys(devDeps || {}).sort()).toEqual(REQUIRED_DEPENDENCIES.sort());

    // report
    const reportPath = resolve(basePath, '.rehearsal');
    expect(readdirSync(reportPath)).toContain('migrate-report.sarif');

    // tsconfig.json
    const tsConfig = readJSONSync(resolve(basePath, 'tsconfig.json')) as TSConfig;
    expect(tsConfig).matchSnapshot();

    // lint config
    expect(fileList).toContain('.eslintrc.js');
    expect(fileList).toContain('.rehearsal-eslintrc.js');
    const lintConfig = readFileSync(resolve(basePath, '.eslintrc.js'), { encoding: 'utf-8' });
    const lintConfigDefualt = readFileSync(resolve(basePath, '.rehearsal-eslintrc.js'), {
      encoding: 'utf-8',
    });
    expect(lintConfig).toMatchSnapshot();
    expect(lintConfigDefualt).toMatchSnapshot();

    // new scripts
    expect(packageJson?.scripts?.['lint:tsc']).toBe('tsc --noEmit');
  });

  test('migrate would skip steps after migrate init', async () => {
    // migrate init
    await runBin('migrate', ['init'], {
      cwd: basePath,
    });

    let fileList = readdirSync(basePath);

    // Dependencies
    const packageJson = JSON.parse(
      await fs.readFile(resolve(basePath, 'package.json'), 'utf-8')
    ) as PackageJson;
    const devDeps = packageJson.devDependencies;
    expect(Object.keys(devDeps || {}).sort()).toEqual(REQUIRED_DEPENDENCIES.sort());

    // tsconfig.json
    const tsConfig = readJSONSync(resolve(basePath, 'tsconfig.json')) as TSConfig;
    expect(tsConfig).matchSnapshot();

    // lint config
    expect(fileList).toContain('.eslintrc.js');
    expect(fileList).toContain('.rehearsal-eslintrc.js');
    const lintConfig = readFileSync(resolve(basePath, '.eslintrc.js'), { encoding: 'utf-8' });
    const lintConfigDefualt = readFileSync(resolve(basePath, '.rehearsal-eslintrc.js'), {
      encoding: 'utf-8',
    });
    expect(lintConfig).toMatchSnapshot();
    expect(lintConfigDefualt).toMatchSnapshot();

    // new scripts
    expect(packageJson?.scripts?.['lint:tsc']).toBe('tsc --noEmit');

    // run migrate
    const { stdout } = await runBin('migrate', ['--ci'], {
      cwd: basePath,
    });
    // migrate init output
    expect(cleanOutput(stdout, basePath)).toMatchSnapshot();

    // read files again
    fileList = readdirSync(basePath);
    // file structures
    expect(fileList).toContain('index.ts');
    expect(fileList).toContain('foo.ts');
    expect(fileList).toContain('depends-on-foo.ts');
    expect(fileList).not.toContain('index.js');
    expect(fileList).not.toContain('foo.js');
    expect(fileList).not.toContain('depends-on-foo.js');

    // file contents
    expect(readFileSync(resolve(basePath, 'foo.ts'), { encoding: 'utf-8' })).toMatchSnapshot();
    expect(
      readFileSync(resolve(basePath, 'depends-on-foo.ts'), { encoding: 'utf-8' })
    ).toMatchSnapshot();
    expect(readFileSync(resolve(basePath, 'index.ts'), { encoding: 'utf-8' })).toMatchSnapshot();

    // report
    const reportPath = resolve(basePath, '.rehearsal');
    expect(readdirSync(reportPath)).toContain('migrate-report.sarif');
  });

  test('--skip-init option', async () => {
    // run migrate with --skip-init
    // this command should fail
    expect.assertions(1);
    try {
      const { stdout } = await runBin('migrate', ['--skip-init', '--ci'], {
        cwd: basePath,
      });
      // validate -> initialize -> analyze -> convert
      const expected = [
        '[STARTED] Validate project',
        '[SUCCESS] Validate project',
        '[STARTED] Initialize',
        '[DATA] Setting up config for initialization',
        '[SUCCESS] Initialize',
        '[STARTED] Analyzing Project',
        '[DATA] Running migration on initialization',
        '[SUCCESS] Analyzing Project',
        '[STARTED] Convert JS files to TS`;',
      ].join('\n');
      expect(stdout).toContain(expected);
    } catch (e) {
      // no-ops
    }
  });

  test('Print debug messages with --verbose', async () => {
    const { stdout } = await runBin('migrate', ['--verbose', '--ci'], {
      cwd: basePath,
    });

    expect(cleanOutput(stdout, basePath)).toMatchSnapshot();
  });

  test('show warning message for missing config with --regen', async () => {
    const { stdout } = await runBin('migrate', ['-r', '--ci'], {
      cwd: basePath,
    });
    expect(stdout).toContain('Eslint config (.eslintrc.{js,yml,json,yaml}) does not exist');
    expect(stdout).toContain('tsconfig.json does not exist');
  });

  test('regen result after the first pass', async () => {
    // first run without --regen
    await runBin('migrate', ['--ci'], {
      cwd: basePath,
    });
    const { stdout } = await runBin('migrate', ['-r', '--ci'], {
      cwd: basePath,
    });
    expect(cleanOutput(stdout, basePath)).toMatchSnapshot();
  });

  describe('user defined options passed by --user-config -u', () => {
    function createUserConfig(basePath: string, config: CustomConfig): void {
      const configPath = resolve(basePath, 'rehearsal-config.json');
      writeJSONSync(configPath, config);
    }

    test('migrate.exclude', async () => {
      const files = getFiles('simple');
      // Add a directory that we don't want to ignore
      files['some-dir'] = { 'index.js': '// I should be excluded ' };
      const basePath = create(files);

      createUserConfig(basePath, {
        migrate: {
          setup: {
            ts: { command: 'touch', args: ['custom-ts-config-script'] },
          },
          exclude: ['some-dir'],
        },
      });

      const result = await runBin('migrate', ['-d', '-u', 'rehearsal-config.json', '--ci'], {
        cwd: basePath,
      });

      const expected = [
        '[STARTED] Initialize',
        '[DATA] Setting up config for my-package',
        '[SUCCESS] Initialize',
        '[STARTED] Analyzing Project',
        '[DATA] Running migration on my-package',
        '[DATA] List of files will be attempted to migrate:',
        '[DATA]  lib/a.js',
        '[DATA] index.js',
        '[SUCCESS] Analyzing Project',
      ].join('\n');

      expect(result.stdout).contains(expected);
    });

    test('migrate.include', async () => {
      const files = getFiles('simple');

      // test is a default ignored directory in Package.ts
      // adding it to the include list should override that value.
      files['test'] = { 'index.js': '// I should be included ' };
      const basePath = create(files);

      createUserConfig(basePath, {
        migrate: {
          setup: {
            ts: { command: 'touch', args: ['custom-ts-config-script'] },
          },
          include: ['test'],
        },
      });

      const result = await runBin('migrate', ['-d', '-u', 'rehearsal-config.json', '--ci'], {
        cwd: basePath,
      });
      const expected = [
        '[STARTED] Initialize',
        '[DATA] Setting up config for my-package',
        '[SUCCESS] Initialize',
        '[STARTED] Analyzing Project',
        '[DATA] Running migration on my-package',
        '[DATA] List of files will be attempted to migrate:',
        '[DATA]  lib/a.js',
        '[DATA] index.js',
        '[DATA] test/index.js',
        '[SUCCESS] Analyzing Project',
      ].join('\n');

      expect(result.stdout).contains(expected);
    });
  });
});
