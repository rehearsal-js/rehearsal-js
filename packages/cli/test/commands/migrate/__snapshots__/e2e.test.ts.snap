// Vitest Snapshot v1

exports[`migrate - validation > pass in a dirty git in multi runs 1`] = `
"info:    @rehearsal/migrate<test-version>
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Running migration on initilization
[SUCCESS] Initialize
[STARTED] Convert JS files to TS
[TITLE] Migration Complete
[TITLE]   0 JS files converted to TS
[TITLE]   0 errors caught by rehearsal
[TITLE]   0 have been fixed by rehearsal
[TITLE]   0 errors need to be fixed manually
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE]     -- 0 eslint errors, with details in the report
[SUCCESS] Migration Complete
[SUCCESS]   0 JS files converted to TS
[SUCCESS]   0 errors caught by rehearsal
[SUCCESS]   0 have been fixed by rehearsal
[SUCCESS]   0 errors need to be fixed manually
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS]     -- 0 eslint errors, with details in the report"
`;

exports[`migrate: e2e > Print debug messages with --verbose 1`] = `
"info:    @rehearsal/migrate<test-version>
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Running migration on basic
[SUCCESS] Initialize
[STARTED] Install dependencies
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[SUCCESS] Add package scripts
[STARTED] Convert JS files to TS
debug:   migration started
debug:   Base path: <tmp-path>
debug:   sourceFiles: [\\"<tmp-path>/foo.js\\",\\"<tmp-path>/depends-on-foo.js\\",\\"<tmp-path>/index.js\\"]
[DATA] git mv /foo.js to /foo.ts
[DATA] git mv /depends-on-foo.js to /depends-on-foo.ts
[DATA] git mv /index.js to /index.ts
debug:   config file: <tmp-path>/tsconfig.json
debug:   fileNames: [\\"<tmp-path>/foo.ts\\",\\"<tmp-path>/depends-on-foo.ts\\",\\"<tmp-path>/index.ts\\"]
[DATA] processing file: /foo.ts
[DATA] processing file: /depends-on-foo.ts
[DATA] processing file: /index.ts
[TITLE] Migration Complete
[TITLE]   3 JS files converted to TS
[TITLE]   2 errors caught by rehearsal
[TITLE]   1 have been fixed by rehearsal
[TITLE]   1 errors need to be fixed manually
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE]     -- 1 eslint errors, with details in the report
[SUCCESS] Migration Complete
[SUCCESS]   3 JS files converted to TS
[SUCCESS]   2 errors caught by rehearsal
[SUCCESS]   1 have been fixed by rehearsal
[SUCCESS]   1 errors need to be fixed manually
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS]     -- 1 eslint errors, with details in the report"
`;

exports[`migrate: e2e > againt specific basePath via --basePath option 1`] = `
"info:    @rehearsal/migrate<test-version>
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Running migration on base
[SUCCESS] Initialize
[STARTED] Install dependencies
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[SUCCESS] Add package scripts
[STARTED] Convert JS files to TS
[DATA] git mv /index.js to /index.ts
[DATA] processing file: /index.ts
[TITLE] Migration Complete
[TITLE]   1 JS file converted to TS
[TITLE]   3 errors caught by rehearsal
[TITLE]   0 have been fixed by rehearsal
[TITLE]   3 errors need to be fixed manually
[TITLE]     -- 2 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE]     -- 1 eslint errors, with details in the report
[SUCCESS] Migration Complete
[SUCCESS]   1 JS file converted to TS
[SUCCESS]   3 errors caught by rehearsal
[SUCCESS]   0 have been fixed by rehearsal
[SUCCESS]   3 errors need to be fixed manually
[SUCCESS]     -- 2 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS]     -- 1 eslint errors, with details in the report"
`;

exports[`migrate: e2e > againt specific basePath via --basePath option 2`] = `
"/* @ts-expect-error @rehearsal TODO TS7006: Parameter 'baz' implicitly has an 'any' type. */
export function run(baz) {
  /* @ts-expect-error @rehearsal TODO TS2304: Cannot find name 'foo'. */
  return foo(baz);
}
"
`;

exports[`migrate: e2e > againt specific basePath via --basePath option 3`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "module": "es2020",
    "moduleResolution": "node",
    "newLine": "lf",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "strict": true,
    "target": "ES2021",
  },
  "include": [
    "index.ts",
  ],
}
`;

exports[`migrate: e2e > againt specific basePath via --basePath option 4`] = `
"module.exports = {
  extends: [\\"./.rehearsal-eslintrc.js\\"],
};
"
`;

exports[`migrate: e2e > againt specific basePath via --basePath option 5`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate: e2e > default migrate command 1`] = `
"info:    @rehearsal/migrate<test-version>
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Running migration on basic
[SUCCESS] Initialize
[STARTED] Install dependencies
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[SUCCESS] Add package scripts
[STARTED] Convert JS files to TS
[DATA] git mv /foo.js to /foo.ts
[DATA] git mv /depends-on-foo.js to /depends-on-foo.ts
[DATA] git mv /index.js to /index.ts
[DATA] processing file: /foo.ts
[DATA] processing file: /depends-on-foo.ts
[DATA] processing file: /index.ts
[TITLE] Migration Complete
[TITLE]   3 JS files converted to TS
[TITLE]   2 errors caught by rehearsal
[TITLE]   1 have been fixed by rehearsal
[TITLE]   1 errors need to be fixed manually
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE]     -- 1 eslint errors, with details in the report
[SUCCESS] Migration Complete
[SUCCESS]   3 JS files converted to TS
[SUCCESS]   2 errors caught by rehearsal
[SUCCESS]   1 have been fixed by rehearsal
[SUCCESS]   1 errors need to be fixed manually
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS]     -- 1 eslint errors, with details in the report"
`;

exports[`migrate: e2e > default migrate command 2`] = `
"export function foo(name = \\"World\\") {
  return \`Hello \${name}\`;
}
"
`;

exports[`migrate: e2e > default migrate command 3`] = `
"import { foo } from \\"./foo\\";

console.log(foo(\\"hello\\"));
"
`;

exports[`migrate: e2e > default migrate command 4`] = `
"import { foo } from \\"./foo\\";

export function run(baz: string | undefined) {
  return foo(baz);
}
"
`;

exports[`migrate: e2e > default migrate command 5`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "module": "es2020",
    "moduleResolution": "node",
    "newLine": "lf",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "strict": true,
    "target": "ES2021",
  },
  "include": [
    "foo.ts",
    "depends-on-foo.ts",
    "index.ts",
  ],
}
`;

exports[`migrate: e2e > default migrate command 6`] = `
"module.exports = {
  extends: [\\"./.rehearsal-eslintrc.js\\"],
};
"
`;

exports[`migrate: e2e > default migrate command 7`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate: e2e > init flag 1`] = `
"info:    @rehearsal/migrate<test-version>
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Running migration on basic
[SUCCESS] Initialize
[STARTED] Install dependencies
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[SUCCESS] Add package scripts"
`;

exports[`migrate: e2e > init flag 2`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "module": "es2020",
    "moduleResolution": "node",
    "newLine": "lf",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "strict": true,
    "target": "ES2021",
  },
  "include": [],
}
`;

exports[`migrate: e2e > init flag 3`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate: e2e > regen result after the first pass 1`] = `
"info:    @rehearsal/migrate<test-version>
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Regenerating report for TS errors and Eslint errors
[DATA] processing file: /foo.ts
[DATA] processing file: /depends-on-foo.ts
[DATA] processing file: /index.ts
[TITLE] Migration Report Generated
[TITLE]   3 ts files have been scanned
[TITLE]   1 errors caught by rehearsal
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE]     -- 1 eslint errors, with details in the report
[SUCCESS] Migration Report Generated
[SUCCESS]   3 ts files have been scanned
[SUCCESS]   1 errors caught by rehearsal
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS]     -- 1 eslint errors, with details in the report"
`;

exports[`migrate: e2e > user defined options passed by --user-config -u > migrate.exclude 1`] = `
"[32minfo[39m:    @rehearsal/migrate 1.0.9-beta
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize -- Dry Run Mode
[DATA] Running migration on my-package
[DATA] List of files will be attempted to migrate:
[DATA]  lib/a.js
[DATA] index.js
[SUCCESS] Initialize -- Dry Run Mode
[STARTED] Install dependencies
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[SUCCESS] Add package scripts"
`;

exports[`migrate: e2e > user defined options passed by --user-config -u > migrate.include 1`] = `
"[32minfo[39m:    @rehearsal/migrate 1.0.9-beta
[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize -- Dry Run Mode
[DATA] Running migration on my-package
[DATA] List of files will be attempted to migrate:
[DATA]  lib/a.js
[DATA] index.js
[DATA] test/index.js
[SUCCESS] Initialize -- Dry Run Mode
[STARTED] Install dependencies
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[SUCCESS] Add package scripts"
`;
